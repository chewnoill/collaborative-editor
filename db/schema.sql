SET statement_timeout = 0;
SET lock_timeout = 0;
SET idle_in_transaction_session_timeout = 0;
SET client_encoding = 'UTF8';
SET standard_conforming_strings = on;
SELECT pg_catalog.set_config('search_path', '', false);
SET check_function_bodies = false;
SET xmloption = content;
SET client_min_messages = warning;
SET row_security = off;

--
-- Name: uuid-ossp; Type: EXTENSION; Schema: -; Owner: -
--

CREATE EXTENSION IF NOT EXISTS "uuid-ossp" WITH SCHEMA public;


--
-- Name: EXTENSION "uuid-ossp"; Type: COMMENT; Schema: -; Owner: -
--

COMMENT ON EXTENSION "uuid-ossp" IS 'generate universally unique identifiers (UUIDs)';


SET default_tablespace = '';

SET default_table_access_method = heap;

--
-- Name: document; Type: TABLE; Schema: public; Owner: -
--

CREATE TABLE public.document (
    id uuid DEFAULT public.uuid_generate_v4() NOT NULL,
    origin text NOT NULL,
    value text NOT NULL,
    web_rtc_key text NOT NULL
);


--
-- Name: schema_migrations; Type: TABLE; Schema: public; Owner: -
--

CREATE TABLE public.schema_migrations (
    version character varying(255) NOT NULL
);


--
-- Name: user_document; Type: TABLE; Schema: public; Owner: -
--

CREATE TABLE public.user_document (
    document_id uuid,
    user_id uuid
);


--
-- Name: users; Type: TABLE; Schema: public; Owner: -
--

CREATE TABLE public.users (
    id uuid DEFAULT public.uuid_generate_v4() NOT NULL,
    name text NOT NULL
);


--
-- Name: document document_pkey; Type: CONSTRAINT; Schema: public; Owner: -
--

ALTER TABLE ONLY public.document
    ADD CONSTRAINT document_pkey PRIMARY KEY (id);


--
-- Name: schema_migrations schema_migrations_pkey; Type: CONSTRAINT; Schema: public; Owner: -
--

ALTER TABLE ONLY public.schema_migrations
    ADD CONSTRAINT schema_migrations_pkey PRIMARY KEY (version);


--
-- Name: user_document user_document_document_id_user_id_key; Type: CONSTRAINT; Schema: public; Owner: -
--

ALTER TABLE ONLY public.user_document
    ADD CONSTRAINT user_document_document_id_user_id_key UNIQUE (document_id, user_id);


--
-- Name: users users_pkey; Type: CONSTRAINT; Schema: public; Owner: -
--

ALTER TABLE ONLY public.users
    ADD CONSTRAINT users_pkey PRIMARY KEY (id);


--
-- Name: user_document user_document_document_id_fkey; Type: FK CONSTRAINT; Schema: public; Owner: -
--

ALTER TABLE ONLY public.user_document
    ADD CONSTRAINT user_document_document_id_fkey FOREIGN KEY (document_id) REFERENCES public.document(id) ON DELETE CASCADE;


--
-- Name: user_document user_document_user_id_fkey; Type: FK CONSTRAINT; Schema: public; Owner: -
--

ALTER TABLE ONLY public.user_document
    ADD CONSTRAINT user_document_user_id_fkey FOREIGN KEY (user_id) REFERENCES public.users(id) ON DELETE CASCADE;


--
-- PostgreSQL database dump complete
--


--
-- Dbmate schema migrations
--

INSERT INTO public.schema_migrations (version) VALUES
    ('20210629194148');
