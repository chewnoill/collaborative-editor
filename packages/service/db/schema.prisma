generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

model document {
  id            String          @id @default(dbgenerated("uuid_generate_v4()")) @db.Uuid
  value         String
  web_rtc_key   String
  origin        Bytes
  user_document user_document[]
}

model schema_migrations {
  version String @id @db.VarChar(255)
}

model user_document {
  document_id String?   @db.Uuid
  user_id     String?   @db.Uuid
  id          String    @id @default(dbgenerated("uuid_generate_v4()")) @db.Uuid
  document    document? @relation(fields: [document_id], references: [id])
  users       users?    @relation(fields: [user_id], references: [id])

  @@unique([document_id, user_id], name: "user_document_document_id_user_id_key")
}

model users {
  id            String          @id @default(dbgenerated("uuid_generate_v4()")) @db.Uuid
  name          String
  user_document user_document[]
}
