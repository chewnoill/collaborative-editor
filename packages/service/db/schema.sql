SET statement_timeout = 0;
SET lock_timeout = 0;
SET idle_in_transaction_session_timeout = 0;
SET client_encoding = 'UTF8';
SET standard_conforming_strings = on;
SELECT pg_catalog.set_config('search_path', '', false);
SET check_function_bodies = false;
SET xmloption = content;
SET client_min_messages = warning;
SET row_security = off;

--
-- Name: postgraphile_watch; Type: SCHEMA; Schema: -; Owner: -
--

CREATE SCHEMA postgraphile_watch;


--
-- Name: pgcrypto; Type: EXTENSION; Schema: -; Owner: -
--

CREATE EXTENSION IF NOT EXISTS pgcrypto WITH SCHEMA public;


--
-- Name: EXTENSION pgcrypto; Type: COMMENT; Schema: -; Owner: -
--

COMMENT ON EXTENSION pgcrypto IS 'cryptographic functions';


--
-- Name: uuid-ossp; Type: EXTENSION; Schema: -; Owner: -
--

CREATE EXTENSION IF NOT EXISTS "uuid-ossp" WITH SCHEMA public;


--
-- Name: EXTENSION "uuid-ossp"; Type: COMMENT; Schema: -; Owner: -
--

COMMENT ON EXTENSION "uuid-ossp" IS 'generate universally unique identifiers (UUIDs)';


--
-- Name: notify_watchers_ddl(); Type: FUNCTION; Schema: postgraphile_watch; Owner: -
--

CREATE FUNCTION postgraphile_watch.notify_watchers_ddl() RETURNS event_trigger
    LANGUAGE plpgsql
    AS $$
begin
  perform pg_notify(
    'postgraphile_watch',
    json_build_object(
      'type',
      'ddl',
      'payload',
      (select json_agg(json_build_object('schema', schema_name, 'command', command_tag)) from pg_event_trigger_ddl_commands() as x)
    )::text
  );
end;
$$;


--
-- Name: notify_watchers_drop(); Type: FUNCTION; Schema: postgraphile_watch; Owner: -
--

CREATE FUNCTION postgraphile_watch.notify_watchers_drop() RETURNS event_trigger
    LANGUAGE plpgsql
    AS $$
begin
  perform pg_notify(
    'postgraphile_watch',
    json_build_object(
      'type',
      'drop',
      'payload',
      (select json_agg(distinct x.schema_name) from pg_event_trigger_dropped_objects() as x)
    )::text
  );
end;
$$;


SET default_tablespace = '';

SET default_table_access_method = heap;

--
-- Name: users; Type: TABLE; Schema: public; Owner: -
--

CREATE TABLE public.users (
    id uuid DEFAULT public.uuid_generate_v4() NOT NULL,
    name text NOT NULL,
    password text DEFAULT ''::text NOT NULL
);


--
-- Name: TABLE users; Type: COMMENT; Schema: public; Owner: -
--

COMMENT ON TABLE public.users IS '
@name users
@omit create,update,delete';


--
-- Name: create_user(text); Type: FUNCTION; Schema: public; Owner: -
--

CREATE FUNCTION public.create_user(name text) RETURNS public.users
    LANGUAGE sql STRICT
    AS $$
  INSERT INTO users
    (name)
    VALUES (name)
    RETURNING *;
$$;


--
-- Name: current_user_id(); Type: FUNCTION; Schema: public; Owner: -
--

CREATE FUNCTION public.current_user_id() RETURNS uuid
    LANGUAGE sql STABLE
    AS $$
  SELECT nullif(current_setting('app.user_id', true), '')::uuid;
$$;


--
-- Name: user_document; Type: TABLE; Schema: public; Owner: -
--

CREATE TABLE public.user_document (
    document_id uuid,
    user_id uuid,
    id uuid DEFAULT public.uuid_generate_v4() NOT NULL
);


--
-- Name: TABLE user_document; Type: COMMENT; Schema: public; Owner: -
--

COMMENT ON TABLE public.user_document IS '
@name user_document
@omit create,update,delete';


--
-- Name: invite_user_to_document(text, uuid); Type: FUNCTION; Schema: public; Owner: -
--

CREATE FUNCTION public.invite_user_to_document(name text, document_id uuid) RETURNS public.user_document
    LANGUAGE sql STRICT
    AS $$
  INSERT INTO user_document
    (user_id, document_id)
    VALUES ((SELECT users.id FROM users where name=name), document_id)
    RETURNING *;
$$;


--
-- Name: me(); Type: FUNCTION; Schema: public; Owner: -
--

CREATE FUNCTION public.me() RETURNS public.users
    LANGUAGE sql STABLE
    AS $$
  SELECT * FROM users WHERE id = current_user_id()
$$;


--
-- Name: document; Type: TABLE; Schema: public; Owner: -
--

CREATE TABLE public.document (
    id uuid DEFAULT public.uuid_generate_v4() NOT NULL,
    value text NOT NULL,
    web_rtc_key text NOT NULL,
    origin bytea NOT NULL,
    latest_update_time timestamp without time zone NOT NULL,
    creator_id uuid NOT NULL
);


--
-- Name: TABLE document; Type: COMMENT; Schema: public; Owner: -
--

COMMENT ON TABLE public.document IS '
@name document
@omit create,update,delete';


--
-- Name: COLUMN document.latest_update_time; Type: COMMENT; Schema: public; Owner: -
--

COMMENT ON COLUMN public.document.latest_update_time IS '
@name latest_update_time
@omit create,update';


--
-- Name: document_updates_queue; Type: TABLE; Schema: public; Owner: -
--

CREATE TABLE public.document_updates_queue (
    id uuid DEFAULT public.uuid_generate_v4() NOT NULL,
    document_id uuid NOT NULL,
    document_update bytea NOT NULL,
    created_at timestamp without time zone DEFAULT now() NOT NULL
);


--
-- Name: TABLE document_updates_queue; Type: COMMENT; Schema: public; Owner: -
--

COMMENT ON TABLE public.document_updates_queue IS '
@name document_updates_queue
@omit';


--
-- Name: schema_migrations; Type: TABLE; Schema: public; Owner: -
--

CREATE TABLE public.schema_migrations (
    version character varying(255) NOT NULL
);


--
-- Name: TABLE schema_migrations; Type: COMMENT; Schema: public; Owner: -
--

COMMENT ON TABLE public.schema_migrations IS '
@name schema_migrations
@omit';


--
-- Name: document document_pkey; Type: CONSTRAINT; Schema: public; Owner: -
--

ALTER TABLE ONLY public.document
    ADD CONSTRAINT document_pkey PRIMARY KEY (id);


--
-- Name: document_updates_queue document_updates_queue_pkey; Type: CONSTRAINT; Schema: public; Owner: -
--

ALTER TABLE ONLY public.document_updates_queue
    ADD CONSTRAINT document_updates_queue_pkey PRIMARY KEY (id);


--
-- Name: schema_migrations schema_migrations_pkey; Type: CONSTRAINT; Schema: public; Owner: -
--

ALTER TABLE ONLY public.schema_migrations
    ADD CONSTRAINT schema_migrations_pkey PRIMARY KEY (version);


--
-- Name: users unique_user_name; Type: CONSTRAINT; Schema: public; Owner: -
--

ALTER TABLE ONLY public.users
    ADD CONSTRAINT unique_user_name UNIQUE (name);


--
-- Name: user_document user_document_document_id_user_id_key; Type: CONSTRAINT; Schema: public; Owner: -
--

ALTER TABLE ONLY public.user_document
    ADD CONSTRAINT user_document_document_id_user_id_key UNIQUE (document_id, user_id);


--
-- Name: user_document user_document_pkey; Type: CONSTRAINT; Schema: public; Owner: -
--

ALTER TABLE ONLY public.user_document
    ADD CONSTRAINT user_document_pkey PRIMARY KEY (id);


--
-- Name: users users_pkey; Type: CONSTRAINT; Schema: public; Owner: -
--

ALTER TABLE ONLY public.users
    ADD CONSTRAINT users_pkey PRIMARY KEY (id);


--
-- Name: document document_creator_id_fkey; Type: FK CONSTRAINT; Schema: public; Owner: -
--

ALTER TABLE ONLY public.document
    ADD CONSTRAINT document_creator_id_fkey FOREIGN KEY (creator_id) REFERENCES public.users(id);


--
-- Name: document_updates_queue document_updates_queue_document_id_fkey; Type: FK CONSTRAINT; Schema: public; Owner: -
--

ALTER TABLE ONLY public.document_updates_queue
    ADD CONSTRAINT document_updates_queue_document_id_fkey FOREIGN KEY (document_id) REFERENCES public.document(id) ON DELETE CASCADE;


--
-- Name: user_document user_document_document_id_fkey; Type: FK CONSTRAINT; Schema: public; Owner: -
--

ALTER TABLE ONLY public.user_document
    ADD CONSTRAINT user_document_document_id_fkey FOREIGN KEY (document_id) REFERENCES public.document(id) ON DELETE CASCADE;


--
-- Name: user_document user_document_user_id_fkey; Type: FK CONSTRAINT; Schema: public; Owner: -
--

ALTER TABLE ONLY public.user_document
    ADD CONSTRAINT user_document_user_id_fkey FOREIGN KEY (user_id) REFERENCES public.users(id) ON DELETE CASCADE;


--
-- Name: document create_document_for_current_user; Type: POLICY; Schema: public; Owner: -
--

CREATE POLICY create_document_for_current_user ON public.document FOR INSERT TO postgraphile_user WITH CHECK ((creator_id = public.current_user_id()));


--
-- Name: document; Type: ROW SECURITY; Schema: public; Owner: -
--

ALTER TABLE public.document ENABLE ROW LEVEL SECURITY;

--
-- Name: user_document invite_to_document_if_allowed; Type: POLICY; Schema: public; Owner: -
--

CREATE POLICY invite_to_document_if_allowed ON public.user_document FOR INSERT TO postgraphile_user WITH CHECK ((document_id IN ( SELECT document.id
   FROM public.document
  WHERE (document.creator_id = public.current_user_id()))));


--
-- Name: document select_document_if_allowed; Type: POLICY; Schema: public; Owner: -
--

CREATE POLICY select_document_if_allowed ON public.document FOR SELECT TO postgraphile_user USING (((creator_id = public.current_user_id()) OR (id IN ( SELECT user_document.document_id
   FROM public.user_document
  WHERE (user_document.user_id = public.current_user_id())))));


--
-- Name: user_document; Type: ROW SECURITY; Schema: public; Owner: -
--

ALTER TABLE public.user_document ENABLE ROW LEVEL SECURITY;

--
-- Name: postgraphile_watch_ddl; Type: EVENT TRIGGER; Schema: -; Owner: -
--

CREATE EVENT TRIGGER postgraphile_watch_ddl ON ddl_command_end
         WHEN TAG IN ('ALTER AGGREGATE', 'ALTER DOMAIN', 'ALTER EXTENSION', 'ALTER FOREIGN TABLE', 'ALTER FUNCTION', 'ALTER POLICY', 'ALTER SCHEMA', 'ALTER TABLE', 'ALTER TYPE', 'ALTER VIEW', 'COMMENT', 'CREATE AGGREGATE', 'CREATE DOMAIN', 'CREATE EXTENSION', 'CREATE FOREIGN TABLE', 'CREATE FUNCTION', 'CREATE INDEX', 'CREATE POLICY', 'CREATE RULE', 'CREATE SCHEMA', 'CREATE TABLE', 'CREATE TABLE AS', 'CREATE VIEW', 'DROP AGGREGATE', 'DROP DOMAIN', 'DROP EXTENSION', 'DROP FOREIGN TABLE', 'DROP FUNCTION', 'DROP INDEX', 'DROP OWNED', 'DROP POLICY', 'DROP RULE', 'DROP SCHEMA', 'DROP TABLE', 'DROP TYPE', 'DROP VIEW', 'GRANT', 'REVOKE', 'SELECT INTO')
   EXECUTE FUNCTION postgraphile_watch.notify_watchers_ddl();


--
-- Name: postgraphile_watch_drop; Type: EVENT TRIGGER; Schema: -; Owner: -
--

CREATE EVENT TRIGGER postgraphile_watch_drop ON sql_drop
   EXECUTE FUNCTION postgraphile_watch.notify_watchers_drop();


--
-- PostgreSQL database dump complete
--


--
-- Dbmate schema migrations
--

INSERT INTO public.schema_migrations (version) VALUES
    ('20210629194148'),
    ('20210713211341'),
    ('20210715214346'),
    ('20210720160437'),
    ('20210722161826'),
    ('20210722181845'),
    ('20210723153150'),
    ('20210726125556'),
    ('20210726224018'),
    ('20210729151156'),
    ('20210802130820');
