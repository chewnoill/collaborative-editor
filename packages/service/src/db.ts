/// <reference types="./zapatos/schema" />

export * as db from "zapatos/db";
import { postgraphile } from "postgraphile";
import type * as schema from "zapatos/schema";
import url from "whatwg-url";
import Pool from "pg-pool";
import { IS_PROD } from "./config";
import DocumentMutations from "./resolvers/documents";
import RandomQueries from "./resolvers/random-names";
import MdxQueries from "./resolvers/render-mdx";
import { DATABASE_URL } from "./env";
export { schema };

const params = url.parseURL(DATABASE_URL);

const config = {
  user: params.username,
  password: params.password,
  host: params.host,
  port: params.port,
  database: params.path[0],
};

export const pool = new Pool(config);

const postgraphile_options: any = {
  graphqlRoute: "/api/graphql",
  pgSettings(req) {
    return {
      role: "postgraphile_user",
      "app.user_id": req.user?.id,
    };
  },
  watchPg: IS_PROD ? false : true,
  graphiql: true,
  enhanceGraphiql: true,
  subscriptions: true,
  dynamicJson: true,
  setofFunctionsContainNulls: false,
  ignoreRBAC: false,
  showErrorStack: "json",
  extendedErrors: ["hint", "detail", "errcode"],
  allowExplain: true,
  legacyRelations: "omit",
  sortExport: true,
  appendPlugins: [DocumentMutations, RandomQueries, MdxQueries],
};

export function gqlMiddleware() {
  return postgraphile(DATABASE_URL, ["public"], postgraphile_options);
}
