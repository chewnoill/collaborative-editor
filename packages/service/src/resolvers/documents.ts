import { makeExtendSchemaPlugin, gql } from "graphile-utils";
import { toBuffer } from "zapatos/db";
import { db } from "../db";
import { queue } from "../mq";
import {
  createDocument,
  insertUpdate,
  updateDocumentMeta,
} from "../utils/documents";

const DocumentMutations = makeExtendSchemaPlugin((build) => {
  return {
    typeDefs: gql`
      extend type Mutation {
        # Individual record
        createDoc(name: String!): Document
        updateDocument(id: UUID!, update: DocumentUpdateInput!): Document
        editDocument(id: UUID!, update: String!): Boolean
      }
      input DocumentUpdateInput {
        isPublic: Boolean
        name: String
      }
    `,
    resolvers: {
      Mutation: {
        createDoc(_, { name }, { pgClient }) {
          return createDocument(pgClient, { name });
        },
        async editDocument(_, { id, update }, { pgClient }) {
          await insertUpdate(id, toBuffer(update), {
            user_id: db.sql`current_user_id()`,
          }).run(pgClient);
          await queue.add("update-document", { document_id: id, update });
          return true;
        },
        async updateDocument(_, { id, update }, { pgClient }) {
          const meta = await updateDocumentMeta(id, {
            name: update.name,
            is_public: update.isPublic,
          }).run(pgClient);
          if (meta.length === 0) return null;
          return {
            ...meta[0],
            isPublic: meta[0].is_public,
            latestUpdateTime: meta[0].latest_update_time,
            creatorId: meta[0].creator_id,
          };
        },
      },
    },
  };
});

export default DocumentMutations;
