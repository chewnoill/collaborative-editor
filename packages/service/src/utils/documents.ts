import { Pool } from "pg";
import * as Y from "yjs";
import { db, pool, schema } from "../db";

const TEXT_NAME = "codemirror";

export function insertUpdate(document_id: string, document_update: Uint8Array) {
  return db
    .insert("document_updates_queue", {
      document_id,
      document_update: Buffer.from(document_update),
    })
    .run(pool);
}

export const createDocument = (pool: Pool, doc: Y.Doc = new Y.Doc()) =>
  db
    .insert("document", {
      value: doc.getText(TEXT_NAME).toJSON(),
      origin: Buffer.from(Y.encodeStateAsUpdate(doc)),
      web_rtc_key: "web_rtc_key",
      creator_id: db.sql`current_user_id()`,
      latest_update_time: db.sql`now()`,
    })
    .run(pool);

export const inviteUser = (
  {
    doc,
    user: current_user,
  }: {
    doc: { id: string };
    user: { id: string };
  },
  user_to_invite: string
) =>
  db
    .insert("user_document", {
      document_id: db.sql`SELECT document.id FROM document WHERE document.creator_id = ${db.param(
        current_user.id
      )} and document.id = ${db.param(doc.id)}`,
      user_id: db.sql`SELECT users.id FROM users WHERE users.name = ${db.param(
        user_to_invite
      )}`,
    })
    .run(pool);

type DocumentWithUpdates = schema.document.Selectable & {
  document_updates?: schema.document_updates_queue.Selectable[];
};

export function fetchDocument(id: string): Promise<DocumentWithUpdates> {
  return db
    .selectExactlyOne(
      "document",
      { id },
      {
        lateral: {
          document_updates: db.select("document_updates_queue", {
            document_id: db.parent("id"),
            created_at: db.conditions.gt(db.parent("latest_update_time")),
          }),
        },
      }
    )
    .run(pool)
    .then((doc: any) => ({
      ...doc,
      latest_update_time: db.toDate(doc.latest_update_time),
      origin: db.toBuffer(doc.origin),
      document_updates: doc.document_updates.map((update) => ({
        ...update,
        created_at: db.toDate(update.created_at),
        document_update: db.toBuffer(update.document_update),
      })),
    }));
}

export const selectUserDocuments = (user: { id: string }) =>
  db.sql`
SELECT ${"document"}.* FROM document
  LEFT JOIN ${"user_document"}
  ON ${"user_document"}.${"document_id"} = ${"document"}.${"id"}
  WHERE ${{ user_id: user.id }}
  OR ${{ creator_id: user.id }}`.run(pool);
