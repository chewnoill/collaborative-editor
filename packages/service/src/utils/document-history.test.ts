import * as Y from "yjs";
import { pool, db } from "../db";
import { createDocument } from "./documents";
import {
  buildDocumentHistoryBuckets,
  getDocumentHistoryFromTable,
  replaceDocumentHistory,
} from "./document-history";

afterAll(() => {
  return pool.end();
});

function insertUpdate(
  document_id: string,
  document_update: Uint8Array,
  {
    user_id,
    created_at = new Date(),
  }: { user_id?: string; created_at?: Date } = {}
) {
  return db
    .insert("document_updates_queue", {
      user_id,
      document_id,
      document_update: Buffer.from(document_update),
      created_at,
    })
    .run(pool);
}

const INTERVAL = 60 * 1000;

async function testFixtures() {
  const user_a = await db.upsert("users", { name: "User A" }, "name").run(pool);
  const user_b = await db.upsert("users", { name: "User B" }, "name").run(pool);

  const ydoc = new Y.Doc();
  const ytext = ydoc.getText("codemirror");
  const { id } = await createDocument(pool, {
    creator_id: user_a.id,
    name: "test document",
    doc: ydoc,
  });

  let origin = Y.encodeStateVector(ydoc);
  ytext.insert(0, "hello world");
  const start_date = new Date();

  await insertUpdate(id, Y.encodeStateAsUpdate(ydoc, origin), {
    user_id: user_b.id,
    created_at: start_date,
  });
  origin = Y.encodeStateVector(ydoc);
  ytext.insert(100, "user b here to write");
  await insertUpdate(id, Y.encodeStateAsUpdate(ydoc, origin), {
    user_id: user_b.id,
    created_at: new Date(start_date.getTime() + INTERVAL),
  });
  origin = Y.encodeStateVector(ydoc);
  ytext.insert(10, "Second Update");
  await insertUpdate(id, Y.encodeStateAsUpdate(ydoc, origin), {
    user_id: user_a.id,
    created_at: new Date(start_date.getTime() + INTERVAL * 2),
  });
  ytext.insert(40, "user a second Update");
  await insertUpdate(id, Y.encodeStateAsUpdate(ydoc, origin), {
    user_id: user_a.id,
    created_at: new Date(start_date.getTime() + INTERVAL * 3),
  });
  ytext.insert(60, "aowejfoaiewjf");
  await insertUpdate(id, Y.encodeStateAsUpdate(ydoc, origin), {
    user_id: user_a.id,
    created_at: new Date(start_date.getTime() + INTERVAL * 4),
  });
  ytext.insert(0, "#YOOOOO\n");
  await insertUpdate(id, Y.encodeStateAsUpdate(ydoc, origin), {
    created_at: new Date(start_date.getTime() + INTERVAL * 5),
  });
  return { id };
}

test("document history can be generated", async () => {
  // create a yjs document A
  const { id } = await testFixtures();
  const id_obj = { id: id };
  const history = await buildDocumentHistoryBuckets(id_obj);

  history.forEach((value, i) =>
    expect({
      ...value,
      user_id: value.user_id || "",
    }).toMatchSnapshot({
      user_id: expect.any(String),
      document_id: expect.any(String),
    })
  );
});

test("document history replaced in table", async () => {
  const { id } = await testFixtures();
  const id_obj = { id: id };
  await replaceDocumentHistory(id_obj);
  const tableHistory = await getDocumentHistoryFromTable(id_obj.id);
  tableHistory.forEach((value, i) =>
    expect({
      ...value,
      user_id: value.user_id || "",
    }).toMatchSnapshot({
      user_id: expect.any(String),
      document_id: expect.any(String),
      id: expect.any(String),
    })
  );
});
