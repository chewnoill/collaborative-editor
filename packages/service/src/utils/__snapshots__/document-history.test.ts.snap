// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`document history can be generated 1`] = `
Object {
  "content": "hello world",
  "diff": "@@ -0,0 +1 @@
+hello world
\\\\ No newline at end of file
",
  "document_id": Any<String>,
  "user_id": Any<String>,
}
`;

exports[`document history can be generated 2`] = `
Object {
  "content": "hello worlduser b here to write",
  "diff": "@@ -1 +1 @@
-hello world
\\\\ No newline at end of file
+hello worlduser b here to write
\\\\ No newline at end of file
",
  "document_id": Any<String>,
  "user_id": Any<String>,
}
`;

exports[`document history can be generated 3`] = `
Object {
  "content": "hello worlSecond Updateduser b here to write",
  "diff": "@@ -1 +1 @@
-hello worlduser b here to write
\\\\ No newline at end of file
+hello worlSecond Updateduser b here to write
\\\\ No newline at end of file
",
  "document_id": Any<String>,
  "user_id": Any<String>,
}
`;

exports[`document history can be generated 4`] = `
Object {
  "content": "hello worlSecond Updateduser b here to wuser a second Updaterite",
  "diff": "@@ -1 +1 @@
-hello worlSecond Updateduser b here to write
\\\\ No newline at end of file
+hello worlSecond Updateduser b here to wuser a second Updaterite
\\\\ No newline at end of file
",
  "document_id": Any<String>,
  "user_id": Any<String>,
}
`;

exports[`document history can be generated 5`] = `
Object {
  "content": "hello worlSecond Updateduser b here to wuser a second Updateaowejfoaiewjfrite",
  "diff": "@@ -1 +1 @@
-hello worlSecond Updateduser b here to wuser a second Updaterite
\\\\ No newline at end of file
+hello worlSecond Updateduser b here to wuser a second Updateaowejfoaiewjfrite
\\\\ No newline at end of file
",
  "document_id": Any<String>,
  "user_id": Any<String>,
}
`;

exports[`document history can be generated 6`] = `
Object {
  "content": "#YOOOOO
hello worlSecond Updateduser b here to wuser a second Updateaowejfoaiewjfrite",
  "diff": "@@ -1 +1,2 @@
+#YOOOOO
 hello worlSecond Updateduser b here to wuser a second Updateaowejfoaiewjfrite
\\\\ No newline at end of file
",
  "document_id": Any<String>,
  "user_id": Any<String>,
}
`;

exports[`document history replaced in table 1`] = `
Object {
  "diff": "@@ -0,0 +1 @@
+hello world
\\\\ No newline at end of file
",
  "document_id": Any<String>,
  "id": Any<String>,
  "user_id": Any<String>,
}
`;

exports[`document history replaced in table 2`] = `
Object {
  "diff": "@@ -1 +1 @@
-hello world
\\\\ No newline at end of file
+hello worlduser b here to write
\\\\ No newline at end of file
",
  "document_id": Any<String>,
  "id": Any<String>,
  "user_id": Any<String>,
}
`;

exports[`document history replaced in table 3`] = `
Object {
  "diff": "@@ -1 +1 @@
-hello worlduser b here to write
\\\\ No newline at end of file
+hello worlSecond Updateduser b here to write
\\\\ No newline at end of file
",
  "document_id": Any<String>,
  "id": Any<String>,
  "user_id": Any<String>,
}
`;

exports[`document history replaced in table 4`] = `
Object {
  "diff": "@@ -1 +1 @@
-hello worlSecond Updateduser b here to write
\\\\ No newline at end of file
+hello worlSecond Updateduser b here to wuser a second Updaterite
\\\\ No newline at end of file
",
  "document_id": Any<String>,
  "id": Any<String>,
  "user_id": Any<String>,
}
`;

exports[`document history replaced in table 5`] = `
Object {
  "diff": "@@ -1 +1 @@
-hello worlSecond Updateduser b here to wuser a second Updaterite
\\\\ No newline at end of file
+hello worlSecond Updateduser b here to wuser a second Updateaowejfoaiewjfrite
\\\\ No newline at end of file
",
  "document_id": Any<String>,
  "id": Any<String>,
  "user_id": Any<String>,
}
`;

exports[`document history replaced in table 6`] = `
Object {
  "diff": "@@ -1 +1,2 @@
+#YOOOOO
 hello worlSecond Updateduser b here to wuser a second Updateaowejfoaiewjfrite
\\\\ No newline at end of file
",
  "document_id": Any<String>,
  "id": Any<String>,
  "user_id": Any<String>,
}
`;
